### PR TITLE
Define features of VK_EXT_robustness2 as a struct in PipelineOptions

### PR DESCRIPTION
**Pull request by xuechen417**
Define ExtendedRobustness struct to hold the features the app enabled.

### CODE REVIEW

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit fe67978e13ab6b9428e25cde10006952a7f0232e
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 82390674c40402c81dbf311a5ee488fb972894f5</li>
<li>SPVGEN: dev 2f679769a8491f423f5cf2f06adfa2870a82935d</li>
<li>XGL: dev 116d28a446eba3a9f8e6f577bab34fc07e090993</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev a7cfa00580cd7aa8e66f254d1e200e976d90b08e</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 964/964 (100.0%)</li>
<li>Failed: 0/964 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx8</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Gfx9</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit 581fd639aa0bf7d10ac91dbdb67ad8995f9efe3b
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 82390674c40402c81dbf311a5ee488fb972894f5</li>
<li>SPVGEN: dev 2f679769a8491f423f5cf2f06adfa2870a82935d</li>
<li>XGL: dev 116d28a446eba3a9f8e6f577bab34fc07e090993</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0c3b9cd01021a1b221ea7b08733989d322adf395</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 964/964 (100.0%)</li>
<li>Failed: 0/964 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx8</summary><ul>

<li>Passed: 103165/103165 (100.0%)</li>
<li>Failed: 0/103165 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Gfx9</summary><ul>

<li>Passed: 103221/103221 (100.0%)</li>
<li>Failed: 0/103221 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx10</summary><ul>

<li>Passed: 103221/103221 (100.0%)</li>
<li>Failed: 0/103221 (0.0%)</li>
</ul></details>


</ul></details>

File: include/vkgcDefs.h [hunk 1]
```diff
@@ -254,6 +254,13 @@ enum class ShadowDescriptorTableUsage : unsigned {
   Disable = 2,
 };
 
+/// Represent the features of VK_EXT_robustness2
+struct ExtendedRobustness {
+  bool robustBufferAccess; ///< robustBufferAccess2
```
Comment (amdrexu) [inline include/vkgcDefs.h:259]: robustBufferAccess2 -> Robust buffer access, ...
Add more contents to explain what the feature is.

File: include/vkgcDefs.h [hunk 2]
```diff
@@ -254,6 +254,13 @@ enum class ShadowDescriptorTableUsage : unsigned {
...
+  bool robustImageAccess;  ///< robustImageAccess2
```
Comment (amdrexu) [inline include/vkgcDefs.h:260]: robustImageAccess2 -> ///< Robust image access, ...

File: include/vkgcDefs.h [hunk 3]
```diff
@@ -254,6 +254,13 @@ enum class ShadowDescriptorTableUsage : unsigned {
...
+  bool nullDescriptor;     ///< nullDescriptor
```
Comment (amdrexu) [inline include/vkgcDefs.h:261]: nullDescriptor -> ///< Null descriptor, ...

File: include/vkgcDefs.h [hunk 4]
```diff
@@ -267,8 +274,8 @@ struct PipelineOptions {
 
   ShadowDescriptorTableUsage shadowDescriptorTableUsage; ///< Controls shadow descriptor table.
   unsigned shadowDescriptorTablePtrHigh;                 ///< Sets high part of VA ptr for shadow descriptor table.
-  bool nullDescriptor;                                   ///< If set, support VK_EXT_robustness2 to give defined
-                                                         ///  behavior for null descriptor
+  ExtendedRobustness extendedRobustness;                 ///< If set, support VK_EXT_robustness2 to give defined
```
Comment (amdrexu) [inline include/vkgcDefs.h:277]: ///< If set, support VK_EXT_robustness2 to give defined behavior for null descriptor and out-of-range access
->
///<  If set, support VK_EXT_robustness2 features

Move the feature explanations to ExtendedRobustness field comments.

File: include/vkgcDefs.h (hunk 1 above)
Comment (xuechen417) [reply include/vkgcDefs.h]: Done.

File: include/vkgcDefs.h (hunk 2 above)
Comment (xuechen417) [reply include/vkgcDefs.h]: Done.

File: include/vkgcDefs.h (hunk 3 above)
Comment (xuechen417) [reply include/vkgcDefs.h]: Done.

File: include/vkgcDefs.h (hunk 4 above)
Comment (xuechen417) [reply include/vkgcDefs.h]: Done.

File: include/vkgcDefs.h [hunk 5]
```diff
@@ -267,8 +277,7 @@ struct PipelineOptions {
 
   ShadowDescriptorTableUsage shadowDescriptorTableUsage; ///< Controls shadow descriptor table.
   unsigned shadowDescriptorTablePtrHigh;                 ///< Sets high part of VA ptr for shadow descriptor table.
-  bool nullDescriptor;                                   ///< If set, support VK_EXT_robustness2 to give defined
-                                                         ///  behavior for null descriptor
+  ExtendedRobustness extendedRobustness;                 ///< If set, support VK_EXT_robustness2 features.
```
Comment (linqun) [inline include/vkgcDefs.h:280]: what's the difference between PipelineOptions::robustBufferAccess and PipelineOptions::ExtendedRobustness::robustBufferAccess? could you use a different name if they have difference?

Comment (xuechen417) [reply include/vkgcDefs.h:280]: PipelineOptions::ExtendedRobustness::robustBufferAccess represents the feature robustBufferAccess2 in VK_EXT_robustness2, which has more sticker requirements of out-of-bounds check than robustBufferAccess core feature.

Comment (nhaehnle) [reply include/vkgcDefs.h:280]: Please adjust the comments slightly to make this clearer.

PipelineOptions::robustBufferAccess should say:

> If set, out of bounds accesses to buffer or private array do not cause a fault, but may read or write data anywhere in the buffer or array.

PipelineOptions::ExtendedRobustness::robustBufferAccess should say:

> Whether buffer accesses are tightly bounds-checked against the range of the descriptor. Give defined behavior (e.g., read 0) for out-of-bounds buffer access and descriptor range rounding up.

By making these adjustments, we give the reader a high-level idea of the distinction. It is okay to refer them to the Vulkan spec for more detail. For example, a comment could say that ExtendedRobustness is intended to correspond to the features of VK_EXT_robustness2.

Comment (xuechen417) [reply include/vkgcDefs.h]: Thanks. Fixed.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
cancel retest because of conflict.

**Comment by xuechen417** [comment]
Rebased and resolved the conficts.

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by JaxLinAMD** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
### Test summary for commit d55777078eb4855d0c68ffc115021fca51af5a5f
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev 82390674c40402c81dbf311a5ee488fb972894f5</li>
<li>SPVGEN: dev 2f679769a8491f423f5cf2f06adfa2870a82935d</li>
<li>XGL: dev 116d28a446eba3a9f8e6f577bab34fc07e090993</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 3274df1841b8f799a490a6be7d58ef9870662fa7</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 964/964 (100.0%)</li>
<li>Failed: 0/964 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 103221/103221 (100.0%)</li>
<li>Failed: 0/103221 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**